### PR TITLE
sjawhar-legion-399: restore FileRegistry and SessionLookup pluggable interface

### LINKED ISSUE #399: fix(envoy): restore FileRegistry and SessionLookup pluggable interface removed by PR #393
## Problem

PR #393 (merged via #373) stripped the file registry entirely, but the design from #375 was a pluggable `SessionLookup` interface with two implementations:
- `KVRegistry` (NATS-backed, for production/multi-machine)
- `FileRegistry` (filesystem-backed, for local dev/single-machine)

Selected by `ENVOY_SESSION_REGISTRY` env var, never both at once.

PR #393 correctly removed:
- ✅ Dual-registry fallback
- ✅ File-writing from plugin when mode=kv
- ✅ ENVOY_REGISTRY_DIR from Pulumi (prod uses KV)

But incorrectly also removed:
- ❌ `FileRegistry` implementation
- ❌ `NewSessionLookup` factory
- ❌ `ENVOY_SESSION_REGISTRY` mode switch logic

This breaks single-machine setups that don't want NATS just for session tracking.

## Fix

Restore from PR #375's implementation:
1. Restore `FileRegistry` struct and its `SessionLookup` interface methods
2. Restore `NewSessionLookup` factory function
3. Restore `ENVOY_SESSION_REGISTRY` env var config and mode switch
4. Keep the KV-only default for production

## References
- PR #375 (SessionLookup interface + registry mode switch)
- PR #393 (over-scoped removal)
- Envoy domain expert review feedback

## Scope
Go code in `packages/envoy/` only.

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #399

## Summary

PR #393 correctly stripped the dual-registry fallback but over-scoped and removed the pluggable `SessionLookup` interface entirely. This restores:

- **FileRegistry** implementation (filesystem-backed `SessionLookup` for local dev)
- **NewSessionLookup** factory function (mode switch via `ENVOY_SESSION_REGISTRY` env var)
- **ParseRegistryMode** (reads `ENVOY_SESSION_REGISTRY`, defaults to `kv`)
- **ENVOY_SESSION_REGISTRY** wiring in listener `main()` — uses interface type, type assertion for `/v1/sessions`
- Full test coverage for all restored code (8 FileRegistry tests, 4 ParseRegistryMode tests, 6 NewSessionLookup tests, 1 nil-sessions-503 listener test)

The KV-only default for production is unchanged. Single-machine setups can use `ENVOY_SESSION_REGISTRY=file` to avoid NATS dependency for session tracking.

**Base:** `sjawhar-legion-373` (PR #393 must merge first)

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 4/4 acceptance criteria verified

### Results
| Criterion | Status | Evidence |
|-----------|--------|----------|
| Restore FileRegistry struct with SessionLookup methods (Get/Put/Delete) | ✅ | `file_registry.go:16-60` — struct + 3 methods. 8 unit tests pass: legacy format, SessionEntry format, not found, zero port, put+get, delete, delete nonexistent, invalid JSON. `go test ./internal/session/ -run TestFileRegistry -v` → 8 PASS |
| Restore NewSessionLookup factory function | ✅ | `lookup.go:25-47` — factory dispatches to FileRegistry or SessionRegistry by mode. 6 tests pass: KV, File, EmptyDefaultsToKV, FileMissingDir, KVMissingConn, InvalidMode. `go test ./internal/session/ -run TestNewSessionLookup -v` → 6 PASS |
| Restore ENVOY_SESSION_REGISTRY env var config and mode switch | ✅ | `lookup.go:49-57` — ParseRegistryMode reads env var. `main.go:416-427` — wires into listener startup, logs mode. 4 ParseRegistryMode tests pass (Default, KV, File, CaseInsensitive). `go test ./internal/session/ -run TestParseRegistryMode -v` → 4 PASS |
| Keep KV-only default for production | ✅ | `lookup.go:53-54` returns `RegistryModeKV` when env unset/empty. `lookup.go:39` treats empty same as "kv" in factory. TestParseRegistryMode_Default and TestNewSessionLookup_EmptyDefaultsToKV confirm behavior. |

**Full test suite:** `go test ./... -count=1` → all 17 packages pass (including integration at 114s). `go vet ./...` → clean. CI `envoy-go` → ✅ passed.

### Test Quality Critique
- **FileRegistry tests (8 tests):** Good coverage of happy paths AND error paths — legacy format compat, invalid JSON, zero port, nonexistent files. No excessive mocking; all tests use real filesystem via `t.TempDir()`.
- **ParseRegistryMode tests (4 tests):** Covers default, explicit kv, file, and case-insensitive. Uses `t.Setenv` properly (test-scoped env mutation).
- **NewSessionLookup tests (6 tests):** Covers all branches including error validation (missing dir, missing conn, invalid mode). KV tests use real NATS testcontainers — not mocked.
- **Nil-sessions-503 test (1 test):** Good edge case — verifies graceful degradation when FileRegistry is active and /v1/sessions is called (returns 503 since FileRegistry has no List()).
- **No excessive mocking.** NATS-dependent tests use testcontainers. Filesystem tests use `t.TempDir()`.
- **P2: Missing integration test for file mode end-to-end.** No test boots the listener with `ENVOY_SESSION_REGISTRY=file` and exercises the full subscribe→deliver path. The unit tests prove the components work individually, but the integration gap means file-mode delivery correctness is unverified at the system level.

### Documentation Feedback
- PR description is clear and well-structured with explicit sub-items for each restored component.
- No README or doc changes in this PR. The existing deploy/README.md still references `ENVOY_REGISTRY_DIR` from the previous architecture — this was noted as stale by a prior test worker and is not new to this PR.

### Observations
- P2: `FileRegistry.Get` (`file_registry.go:24-26`) wraps **any** `os.ReadFile` error as `os.ErrNotExist`, masking permission or I/O errors as "not found." Not an acceptance failure, but a real edge-case bug.
- P3: `/v1/sessions` returns 503 in file mode (intentional per design — FileRegistry has no `List()`). This is documented in the plan and matches the claimed type-assertion approach.

### Config Compliance
- require_specific_task: false — N/A (default)
- require_taiga_evidence: false — N/A (default)

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 1 issue
**MINOR (P3):** 2 suggestions

**Verdict: Approved to merge** — no blocking issues.

---

### Acceptance Criteria Verification

All 4 requirements from issue #399 are fully addressed:

| Requirement | Status | Evidence |
|-------------|--------|----------|
| Restore `FileRegistry` struct + `SessionLookup` methods | ✅ Met | `file_registry.go`: Get/Put/Delete with legacy format compat, MachineID stamping |
| Restore `NewSessionLookup` factory function | ✅ Met | `lookup.go`: mode-based factory with validation (file requires dir, kv requires conn) |
| Restore `ENVOY_SESSION_REGISTRY` env var + mode switch | ✅ Met | `ParseRegistryMode()` in lookup.go, wired in `main.go` |
| Keep KV-only default for production | ✅ Met | `ParseRegistryMode()` returns `"kv"` when unset/empty |

### Silent Shortcuts Check: None found
- No TODOs, no hardcoded values, no simplified implementations vs PR #375 original

### Key Findings

**P2: Error masking in FileRegistry.Get (file_registry.go:24-25)**
`os.ReadFile` errors (permission denied, IO errors) are unconditionally wrapped as `os.ErrNotExist`. This flattens distinct failure modes into "not found". Tester also flagged this. Acceptable for restoration PR since it matches original PR #375 design — recommend addressing in a follow-up.

**P3: No path traversal guard on sessionID in FileRegistry**
`filepath.Join(f.Dir, sessionID+".json")` — if sessionID contains `../`, the path escapes the registry dir. Low risk: FileRegistry is local-dev only and the listener is typically not internet-exposed. Still, a `filepath.Rel` check or `strings.ContainsAny(sessionID, "/\\")` guard would harden it.

**P3: Type assertion readability (main.go:324)**
`reg, _ := d.sessions.(*session.SessionRegistry)` — the discarded bool is intentional (nil reg → 503 via sessionsHandler), and `TestSessionsHandler_NilSessions_Returns503` documents the behavior. An inline comment would help future readers.

### Test Coverage
19 new tests total:
- 8 FileRegistry unit tests (legacy format, SessionEntry format, not found, zero port, put+get, delete, delete nonexistent, invalid JSON)
- 4 ParseRegistryMode tests (default, kv, file, case insensitive)
- 6 NewSessionLookup tests (kv, file, empty default, missing dir, missing conn, invalid mode)
- 1 nil-sessions-503 handler test

### Evidence
- CI status: all 6 checks pass (contracts, envoy-go, envoy-plugin, lint, test, typecheck)
- Bot review comments: none
- Tester confirmation: 4/4 criteria pass

File: packages/envoy/internal/session/file_registry.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+type FileRegistry struct {
+	Dir       string
+	MachineID string // stamped on returned entries (file registry is local-only)
+}
+
+func (f *FileRegistry) Get(sessionID string) (SessionEntry, error) {
+	path := filepath.Join(f.Dir, sessionID+".json")
+	buf, err := os.ReadFile(path)
+	if err != nil {
+		return SessionEntry{}, fmt.Errorf("session %s: %w", sessionID, os.ErrNotExist)
```
Comment (sjawhar) [inline packages/envoy/internal/session/file_registry.go:25]: **[P2 — IMPORTANT]:** Error masking — `os.ReadFile` failures (permission denied, IO errors) are unconditionally wrapped as `os.ErrNotExist`, flattening distinct failure modes into "not found". Tester also flagged this.

Suggested fix:
```go
if os.IsNotExist(err) {
    return SessionEntry{}, fmt.Errorf("session %s: %w", sessionID, os.ErrNotExist)
}
return SessionEntry{}, fmt.Errorf("session %s: read error: %w", sessionID, err)
```

Acceptable for this restoration PR since it matches the original PR #375 code. Recommend addressing in a follow-up.

File: packages/envoy/internal/session/file_registry.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+//
+// This is the "file" mode of SessionLookup — for single-machine, local dev.
+// The plugin writes files directly; the listener reads them.
+type FileRegistry struct {
+	Dir       string
+	MachineID string // stamped on returned entries (file registry is local-only)
+}
+
+func (f *FileRegistry) Get(sessionID string) (SessionEntry, error) {
+	path := filepath.Join(f.Dir, sessionID+".json")
```
Comment (sjawhar) [inline packages/envoy/internal/session/file_registry.go:22]: **[P3 — MINOR]:** No path traversal guard on `sessionID`. If `sessionID` contains `../`, `filepath.Join` resolves it and the path escapes the registry dir. Low risk since FileRegistry is local-dev only and the listener is typically not internet-exposed. A `strings.ContainsAny(sessionID, "/\\")` guard or `filepath.Rel` check would harden this.

File: packages/envoy/cmd/listener/main.go [hunk 3]
```diff
@@ -321,7 +321,8 @@ func main() {
 	})
 	v1.HandleFunc("/v1/sessions", func(w http.ResponseWriter, r *http.Request) {
 		d := deps.Load()
-		sessionsHandler(d.registry, d.sessions).ServeHTTP(w, r)
+		reg, _ := d.sessions.(*session.SessionRegistry)
```
Comment (sjawhar) [inline packages/envoy/cmd/listener/main.go:324]: **[P3 — MINOR]:** The discarded bool from the type assertion is intentional design (nil reg → 503 via `sessionsHandler`), and `TestSessionsHandler_NilSessions_Returns503` documents this. An inline comment like `// nil when mode=file; sessionsHandler returns 503` would help future readers.
